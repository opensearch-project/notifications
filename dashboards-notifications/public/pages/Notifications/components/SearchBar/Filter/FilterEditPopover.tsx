/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useContext, useMemo, useState } from 'react';
import { MainContext } from '../../../../Main/Main';
import {
  FilterOperatorType,
  FilterFieldType,
  getFilterOperatorOptions,
  getOperatorString,
  getFilterFieldOptions,
  getValueComponent,
} from '../utils/filterHelpers';
import { FilterType } from './Filters';

interface FilterEditPopoverProps {
  filter?: FilterType;
  closePopover: () => void;
  index?: number;
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function FilterEditPopover(props: FilterEditPopoverProps) {
  const mainStateContext = useContext(MainContext)!;
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [
          {
            label: getOperatorString(
              props.filter.operator,
              props.filter.inverted
            ),
          },
        ]
      : []
  );
  const [filterValue, setFilterValue] = useState<FilterType['value']>(
    props.filter?.value || []
  );

  const filterFieldOptions = useMemo(getFilterFieldOptions, []);

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button
        style={{
          width: 0,
          height: 0,
          position: 'fixed',
          marginLeft: -10000,
          bottom: 0,
        }}
      />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label="Field">
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label="Operator">
            <EuiComboBox
              placeholder={
                selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'
              }
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getFilterOperatorOptions(
                      selectedFieldOptions[0].label as FilterFieldType
                    )
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label as FilterFieldType,
          selectedOperatorOptions[0].label as FilterOperatorType,
          filterValue,
          setFilterValue,
          mainStateContext.availableChannels
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              !(filterValue && filterValue.length > 0)
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label as FilterFieldType,
                  operator: selectedOperatorOptions[0]
                    .label as FilterOperatorType,
                  value: filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
