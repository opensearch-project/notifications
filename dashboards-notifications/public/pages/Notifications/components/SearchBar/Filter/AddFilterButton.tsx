/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButtonEmpty, EuiPopover, EuiPopoverTitle } from '@elastic/eui';
import React, { useState } from 'react';
import { FilterEditPopover } from './FilterEditPopover';
import { FilterType } from './Filters';

interface AddFilterButtonProps {
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function AddFilterButton(props: AddFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const button = (
    <EuiButtonEmpty
      size="xs"
      onClick={() => {
        setIsPopoverOpen(true);
      }}
    >
      + Add filter
    </EuiButtonEmpty>
  );

  return (
    <>
      <EuiPopover
        button={button}
        isOpen={isPopoverOpen}
        closePopover={() => setIsPopoverOpen(false)}
        anchorPosition="downLeft"
        withTitle
      >
        <EuiPopoverTitle>{'Add filter'}</EuiPopoverTitle>
        <FilterEditPopover
          setFilter={props.setFilter}
          closePopover={() => setIsPopoverOpen(false)}
        />
      </EuiPopover>
    </>
  );
}
