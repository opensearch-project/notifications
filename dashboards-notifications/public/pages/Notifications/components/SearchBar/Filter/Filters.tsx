/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiBadge,
  EuiContextMenu,
  EuiContextMenuPanelDescriptor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiPopover,
  EuiTextColor,
} from '@elastic/eui';
import { EuiComboBoxOptionOption } from '@elastic/eui/src/components/combo_box/types';
import React, { useMemo, useState } from 'react';
import {
  FilterFieldType,
  FilterOperatorType,
  filterToQueryString,
  isSingleSelection,
} from '../utils/filterHelpers';
import { AddFilterButton } from './AddFilterButton';
import { FilterEditPopover } from './FilterEditPopover';
import { GlobalFilterButton } from './GlobalFilterButton';

export interface FilterType {
  field: FilterFieldType;
  operator: FilterOperatorType;
  value: string | Array<EuiComboBoxOptionOption<string>> | null;
  inverted: boolean;
  disabled: boolean;
}

interface FiltersProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function Filters(props: FiltersProps) {
  // set a filter at an index. if newFilter doesn't exist, remove filter at the index
  // if index doesn't exist, append newFilter to the end
  const setFilter = (newFilter: FilterType | null, index?: number) => {
    if (index === undefined) index = props.filters.length;
    const newFilters = [...props.filters];
    if (newFilter) newFilters.splice(index, 1, newFilter);
    else newFilters.splice(index, 1);
    props.setFilters(newFilters);
  };

  const getFilterPopoverPanels = (
    filter: FilterType,
    index: number,
    closePopover: () => void
  ): EuiContextMenuPanelDescriptor[] => [
    {
      id: 0,
      items: [
        {
          name: 'Edit filter',
          icon: <EuiIcon type="invert" size="m" />,
          panel: 1,
        },
        {
          name: `${filter.inverted ? 'Include' : 'Exclude'} results`,
          icon: (
            <EuiIcon
              type={filter.inverted ? 'plusInCircle' : 'minusInCircle'}
              size="m"
            />
          ),
          onClick: () => {
            filter.inverted = !filter.inverted;
            setFilter(filter, index);
          },
        },
        {
          name: filter.disabled ? 'Re-enable' : 'Temporarily disable',
          icon: (
            <EuiIcon type={filter.disabled ? 'eye' : 'eyeClosed'} size="m" />
          ),
          onClick: () => {
            filter.disabled = !filter.disabled;
            setFilter(filter, index);
          },
        },
        {
          name: 'Delete',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => setFilter(null, index),
        },
      ],
    },
    {
      id: 1,
      width: 430,
      title: 'Edit filter',
      content: (
        <div style={{ margin: 15 }}>
          <FilterEditPopover
            filter={filter}
            index={index}
            setFilter={setFilter}
            closePopover={closePopover}
          />
        </div>
      ),
    },
  ];

  const renderFilters = () => {
    const FilterBadge = ({
      filter,
      index,
    }: {
      filter: FilterType;
      index: number;
    }) => {
      const [isPopoverOpen, setIsPopoverOpen] = useState(false);
      const className =
        'globalFilterItem' +
        (filter.disabled ? ' globalFilterItem-isDisabled' : '') +
        (filter.inverted ? ' globalFilterItem-isExcluded' : '');

      let singleSelection = isSingleSelection(filter);
      let value;
      if (typeof filter.value === 'string') {
        value = filter.value;
      } else if (Array.isArray(filter.value)) {
        // combo box array
        value = filter.value.map((option) => option.label).join(', ');
        if (!singleSelection) value = 'is one of ' + value;
      }

      const filterLabel = filter.inverted ? (
        <>
          <EuiTextColor color={filter.disabled ? 'default' : 'danger'}>
            {'NOT '}
          </EuiTextColor>
          <EuiTextColor color="default">{`${filter.field}: ${value}`}</EuiTextColor>
        </>
      ) : (
        `${filter.field}: ${value}`
      );

      const badge = (
        <EuiBadge
          className={className}
          onClick={() => setIsPopoverOpen(true)}
          onClickAriaLabel="Open filter settings"
          color={filter.disabled ? '#e7e9f0' : 'hollow'}
          iconType="cross"
          iconSide="right"
          iconOnClick={() => setFilter(null, index)}
          iconOnClickAriaLabel="Remove filter"
        >
          {filterLabel}
        </EuiBadge>
      );
      return (
        <EuiFlexItem grow={false} key={`filter-${index}`}>
          <EuiPopover
            isOpen={isPopoverOpen}
            closePopover={() => setIsPopoverOpen(false)}
            panelPaddingSize="none"
            button={badge}
          >
            <EuiContextMenu
              initialPanelId={0}
              panels={getFilterPopoverPanels(filter, index, () =>
                setIsPopoverOpen(false)
              )}
            />
          </EuiPopover>
        </EuiFlexItem>
      );
    };

    return (
      <>
        {props.filters.length > 0
          ? props.filters.map((filter, i) => (
              <FilterBadge filter={filter} index={i} key={i} />
            ))
          : null}
      </>
    );
  };

  const filterComponents = useMemo(() => renderFilters(), [props.filters]);

  return (
    <>
      <EuiFlexGroup
        gutterSize="xs"
        alignItems="center"
        responsive={false}
        style={{ minHeight: 32 }}
      >
        <EuiFlexItem grow={false}>
          <GlobalFilterButton
            filters={props.filters}
            setFilters={props.setFilters}
          />
        </EuiFlexItem>
        {filterComponents}
        <EuiFlexItem grow={false}>
          <AddFilterButton setFilter={setFilter} />
        </EuiFlexItem>
      </EuiFlexGroup>
    </>
  );
}
