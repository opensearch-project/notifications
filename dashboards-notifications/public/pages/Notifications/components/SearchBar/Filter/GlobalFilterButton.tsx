/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonIcon,
  EuiContextMenu,
  EuiIcon,
  EuiPopover,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './Filters';

interface GlobalFilterButtonProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function GlobalFilterButton(props: GlobalFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const globalPopoverPanels = [
    {
      id: 0,
      title: 'Change all filters',
      items: [
        {
          name: 'Enable all',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: false }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Disable all',
          icon: <EuiIcon type="eyeClosed" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: true }))
            );
            setIsPopoverOpen(false);
          },
        },
        // {
        //   name: 'Invert inclusion',
        //   icon: <EuiIcon type="invert" size="m" />,
        //   onClick: () => {
        //     props.setFilters(
        //       props.filters.map((filter) => ({
        //         ...filter,
        //         inverted: !filter.inverted,
        //       }))
        //     );
        //     setIsPopoverOpen(false);
        //   },
        // },
        {
          name: 'Invert enabled/disabled',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: !filter.disabled,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Remove all',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => {
            props.setFilters([]);
            setIsPopoverOpen(false);
          },
        },
      ],
    },
  ];

  return (
    <EuiPopover
      isOpen={isPopoverOpen}
      closePopover={() => setIsPopoverOpen(false)}
      button={
        <EuiButtonIcon
          onClick={() => setIsPopoverOpen(true)}
          iconType="filter"
          title="Change all filters"
          aria-label="Change all filters"
        />
      }
      anchorPosition="rightUp"
      panelPaddingSize="none"
      withTitle
    >
      <EuiContextMenu initialPanelId={0} panels={globalPopoverPanels} />
    </EuiPopover>
  );
}
