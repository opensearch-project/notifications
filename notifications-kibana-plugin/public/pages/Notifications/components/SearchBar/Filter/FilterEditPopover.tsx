/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useMemo, useState } from 'react';
import {
  FilterOperatorType,
  FilterFieldType,
  getFilterOperatorOptions,
  getOperatorString,
  getFilterFieldOptions,
  getValueComponent,
} from '../utils/filterHelpers';
import { FilterType } from './Filters';

interface FilterEditPopoverProps {
  filter?: FilterType;
  closePopover: () => void;
  index?: number;
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function FilterEditPopover(props: FilterEditPopoverProps) {
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [
          {
            label: getOperatorString(
              props.filter.operator,
              props.filter.inverted
            ),
          },
        ]
      : []
  );
  const [filterValue, setFilterValue] = useState<FilterType['value']>(
    props.filter?.value || []
  );

  const filterFieldOptions = useMemo(getFilterFieldOptions, []);

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button
        style={{
          width: 0,
          height: 0,
          position: 'fixed',
          marginLeft: -10000,
          bottom: 0,
        }}
      />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label="Field">
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label="Operator">
            <EuiComboBox
              placeholder={
                selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'
              }
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getFilterOperatorOptions(
                      selectedFieldOptions[0].label as FilterFieldType
                    )
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label as FilterFieldType,
          selectedOperatorOptions[0].label as FilterOperatorType,
          filterValue,
          setFilterValue
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              !(filterValue && filterValue.length > 0)
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label as FilterFieldType,
                  operator: selectedOperatorOptions[0]
                    .label as FilterOperatorType,
                  value: filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
