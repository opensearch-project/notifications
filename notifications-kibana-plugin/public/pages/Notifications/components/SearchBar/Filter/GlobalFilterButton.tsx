/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButtonIcon,
  EuiContextMenu,
  EuiIcon,
  EuiPopover,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './Filters';

interface GlobalFilterButtonProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function GlobalFilterButton(props: GlobalFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const globalPopoverPanels = [
    {
      id: 0,
      title: 'Change all filters',
      items: [
        {
          name: 'Enable all',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: false }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Disable all',
          icon: <EuiIcon type="eyeClosed" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: true }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Invert inclusion',
          icon: <EuiIcon type="invert" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                inverted: !filter.inverted,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Invert enabled/disabled',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: !filter.disabled,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Remove all',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => {
            props.setFilters([]);
            setIsPopoverOpen(false);
          },
        },
      ],
    },
  ];

  return (
    <EuiPopover
      isOpen={isPopoverOpen}
      closePopover={() => setIsPopoverOpen(false)}
      button={
        <EuiButtonIcon
          onClick={() => setIsPopoverOpen(true)}
          iconType="filter"
          title="Change all filters"
          aria-label="Change all filters"
        />
      }
      anchorPosition="rightUp"
      panelPaddingSize="none"
      withTitle
    >
      <EuiContextMenu initialPanelId={0} panels={globalPopoverPanels} />
    </EuiPopover>
  );
}
