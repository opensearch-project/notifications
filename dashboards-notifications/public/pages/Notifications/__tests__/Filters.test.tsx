/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { render } from '@testing-library/react';
import React from 'react';
import { Filters, FilterType } from '../component/SearchBar/Filter/Filters';

describe('<Filters /> spec', () => {
  it('renders the component', () => {
    const setFilters = jest.fn();
    const filters: FilterType[] = [
      {
        field: 'Channel',
        operator: 'is',
        value: 'test',
        inverted: false,
        disabled: true,
      },
      {
        field: 'Source',
        operator: 'is one of',
        value: [{ label: 'Alerting' }, { label: 'ISM' }],
        inverted: true,
        disabled: false,
      },
    ];
    const utils = render(<Filters filters={filters} setFilters={setFilters} />);
    expect(utils.container.firstChild).toMatchSnapshot();

    utils.getByText('Channel: test').click();
    utils.getByText('Re-enable').click();
    utils.getByText('Channel: test').click();
    utils.getByText('Exclude results').click();
    utils.getByText('Channel: test').click();
    utils.getByText('Delete').click();
  });
});
