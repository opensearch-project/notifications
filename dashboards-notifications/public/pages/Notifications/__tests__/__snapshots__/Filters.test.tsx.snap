// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> spec renders the component 1`] = `
<div
  class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--wrap"
  style="min-height: 32px;"
>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiPopover euiPopover--anchorRightUp euiPopover--withTitle"
    >
      <div
        class="euiPopover__anchor"
      >
        <button
          aria-label="Change all filters"
          class="euiButtonIcon euiButtonIcon--primary"
          title="Change all filters"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiPopover euiPopover--anchorDownCenter"
    >
      <div
        class="euiPopover__anchor"
      >
        <span
          class="euiBadge euiBadge--iconRight globalFilterItem globalFilterItem-isDisabled"
          style="background-color: rgb(231, 233, 240); color: rgb(0, 0, 0);"
        >
          <span
            class="euiBadge__content"
          >
            <button
              aria-label="Open filter settings"
              class="euiBadge__childButton"
              title="Channel: test"
            >
              Channel: test
            </button>
            <button
              aria-label="Remove filter"
              class="euiBadge__iconButton"
              title="Remove filter"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--small euiIcon-isLoading euiBadge__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiPopover euiPopover--anchorDownCenter"
    >
      <div
        class="euiPopover__anchor"
      >
        <span
          class="euiBadge euiBadge--hollow euiBadge--iconRight globalFilterItem globalFilterItem-isExcluded"
        >
          <span
            class="euiBadge__content"
          >
            <button
              aria-label="Open filter settings"
              class="euiBadge__childButton"
              title="NOT Source: is one of Alerting, ISM"
            >
              <span
                class="euiTextColor euiTextColor--danger"
              >
                NOT 
              </span>
              <span
                class="euiTextColor euiTextColor--default"
              >
                Source: is one of Alerting, ISM
              </span>
            </button>
            <button
              aria-label="Remove filter"
              class="euiBadge__iconButton"
              title="Remove filter"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--small euiIcon-isLoading euiBadge__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiPopover euiPopover--anchorDownLeft euiPopover--withTitle"
    >
      <div
        class="euiPopover__anchor"
      >
        <button
          class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall"
          type="button"
        >
          <span
            class="euiButtonContent euiButtonEmpty__content"
          >
            <span
              class="euiButtonEmpty__text"
            >
              + Add filter
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
